File: apps/keyboard-android/app/src/main/java/com/intelliai/keyboard/dictation/DictationLanguage.kt
package com.intelliai.keyboard.dictation

/**
 * The dictation languages the user can choose — the single canonical
 * mapping between a UI choice and what IntelliAI STT is told.
 *
 * Pure of Android types so the mapping is unit-tested on the JVM;
 * friendly display names live in string resources (a UI concern), and
 * only these four values can ever reach the API — arbitrary language
 * strings never travel through the app.
 *
 *     Auto    → apiTag null → the `language` field is OMITTED (server detects)
 *     English → "en"
 *     Hindi   → "hi"
 *     Arabic  → "ar"
 *
 * ``apiTag`` uses the platform's canonical base subtags (the server
 * normalizes anyway); ``indicator`` is the compact keyboard chip label.
 */
enum class DictationLanguage(
    val apiTag: String?,
    val prefValue: String,
    val indicator: String,
) {
    AUTO(apiTag = null, prefValue = "auto", indicator = "Auto"),
    ENGLISH(apiTag = "en", prefValue = "en", indicator = "EN"),
    HINDI(apiTag = "hi", prefValue = "hi", indicator = "HI"),
    ARABIC(apiTag = "ar", prefValue = "ar", indicator = "AR");

    companion object {
        val DEFAULT = AUTO

        /** Resolve a persisted preference back to a choice; unknown or
         *  absent values fall back to the default (Auto) so a new
         *  install — or a corrupted value — is always Auto. */
        fun fromPref(value: String?): DictationLanguage =
            entries.firstOrNull { it.prefValue == value } ?: DEFAULT
    }
}

File: apps/keyboard-ios/Shared/DictationLanguage.swift
import Foundation

/// The dictation languages the user can choose — the single canonical
/// mapping between a UI choice and what IntelliAI STT is told. Mirrors
/// the Android client and the public API contract exactly:
///
///     Auto    → apiTag nil → the `language` field is OMITTED (server detects)
///     English → "en"
///     Hindi   → "hi"
///     Arabic  → "ar"
///
/// Only these four values can ever reach the API — arbitrary language
/// strings never travel through the app.
enum DictationLanguage: String, CaseIterable, Codable {
    case auto
    case english = "en"
    case hindi = "hi"
    case arabic = "ar"

    static let `default`: DictationLanguage = .auto

    /// What goes into the request's `language` field; nil means omit.
    var apiTag: String? {
        self == .auto ? nil : rawValue
    }

    /// Compact keyboard chip label.
    var indicator: String {
        switch self {
        case .auto: return "Auto"
        case .english: return "EN"
        case .hindi: return "HI"
        case .arabic: return "AR"
        }
    }

    /// Human name for the settings screen.
    var displayName: String {
        switch self {
        case .auto: return "Auto-detect"
        case .english: return "English"
        case .hindi: return "हिन्दी (Hindi)"
        case .arabic: return "العربية (Arabic)"
        }
    }

    /// Resolve a persisted preference back to a choice; unknown or
    /// absent values fall back to Auto so a new install — or a
    /// corrupted value — is always Auto.
    static func fromPreference(_ value: String?) -> DictationLanguage {
        guard let value else { return .default }
        return DictationLanguage(rawValue: value) ?? .default
    }
}
